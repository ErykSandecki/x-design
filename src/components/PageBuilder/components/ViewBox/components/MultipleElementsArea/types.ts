// types
import { TElement } from 'types';

export type TCoordinatesData = {
  elementsCoordinates: Array<{
    coordinates: TRectCoordinates;
    id: TElement['id'];
  }>;
  outline: TRectCoordinates;
};
