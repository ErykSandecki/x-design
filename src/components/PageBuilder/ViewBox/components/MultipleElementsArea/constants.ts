// others
import { BASE_RECT } from 'shared';

// types
import { TCoordinatesData } from './types';

export const INITIAL_COORDINATES_DATA: TCoordinatesData = {
  elementsCoordinates: [],
  outline: BASE_RECT,
};
