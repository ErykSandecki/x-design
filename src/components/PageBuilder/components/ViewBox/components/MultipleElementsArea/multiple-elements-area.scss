@import '../../../../styles/z-indexes';

.MultipleElementsArea {
  &__shadow-element {
    position: absolute;
    z-index: $--xd-zindex-multiple-area;
  }
}
