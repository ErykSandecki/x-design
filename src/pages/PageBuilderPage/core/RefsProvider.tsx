import { createContext, useRef, useContext, FC, ReactNode, useMemo } from 'react';

// types
import { TContext } from './types';

const RefsContext: React.Context<TContext> = createContext(null);

export const useRefs = (): TContext => useContext(RefsContext);

export type TRefsProvider = {
  children: ReactNode;
  itemsRefs?: TContext['itemsRefs'];
  overlayContainerRefHtml?: HTMLDivElement;
  zoomBoxRefHtml?: HTMLDivElement;
  zoomContentRefHtml?: HTMLDivElement;
};

export const RefsProvider: FC<TRefsProvider> = ({
  children,
  itemsRefs: itemRefsDefault = {},
  overlayContainerRefHtml = null,
  zoomBoxRefHtml = null,
  zoomContentRefHtml = null,
}) => {
  const itemRefs = useRef(itemRefsDefault);
  const overlayContainerRef = useRef(overlayContainerRefHtml);
  const zoomBoxRef = useRef(zoomBoxRefHtml);
  const zoomContentRef = useRef(zoomContentRefHtml);
  const value = useMemo(
    () =>
      ({
        itemsRefs: itemRefs.current,
        overlayContainerRef: overlayContainerRef,
        zoomBoxRef: zoomBoxRef,
        zoomContentRef: zoomContentRef,
      }) as TContext,
    [],
  );

  return <RefsContext.Provider value={value}>{children}</RefsContext.Provider>;
};
