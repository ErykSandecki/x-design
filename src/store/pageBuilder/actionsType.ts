export const REDUCER_KEY = 'pageBuilder';

export const ADD_ELEMENT = `${REDUCER_KEY}/ADD_ELEMENT`;
export const APPLY_ELEMENTS_SIZE_TYPE = `${REDUCER_KEY}/APPLY_ELEMENTS_SIZE_TYPE`;
export const CHANGE_ALIGNMENT = `${REDUCER_KEY}/CHANGE_ALIGNMENT`;
export const CHANGE_BACKGROUND = `${REDUCER_KEY}/CHANGE_BACKGROUND`;
export const CHANGE_LAYOUT = `${REDUCER_KEY}/CHANGE_LAYOUT`;
export const CHANGE_PARENT = `${REDUCER_KEY}/CHANGE_PARENT`;
export const CHANGE_POSITION = `${REDUCER_KEY}/CHANGE_POSITION`;
export const CLEAR_PREV_STATE = `${REDUCER_KEY}/CLEAR_PREV_STATE`;
export const FIT_LAYOUT = `${REDUCER_KEY}/FIT_LAYOUT`;
export const FLIP_ELEMENTS = `${REDUCER_KEY}/FLIP_ELEMENTS`;
export const REDUCER_HISTORY_REDO = `${REDUCER_KEY}/REDUCER_HISTORY_REDO`;
export const REDUCER_HISTORY_SAVE = `${REDUCER_KEY}/REDUCER_HISTORY_SAVE`;
export const REDUCER_HISTORY_UNDO = `${REDUCER_KEY}/REDUCER_HISTORY_UNDO`;
export const RESIZE_ELEMENT = `${REDUCER_KEY}/RESIZE_ELEMENT`;
export const ROTATE_ELEMENTS = `${REDUCER_KEY}/ROTATE_ELEMENTS`;
export const SELECT_ELEMENT = `${REDUCER_KEY}/SELECT_ELEMENT`;
export const SELECT_ELEMENTS = `${REDUCER_KEY}/SELECT_ELEMENTS`;
export const SET_AREA_COORDINATES = `${REDUCER_KEY}/SET_AREA_COORDINATES`;
export const SET_ELEMENTS_COORDINATES = `${REDUCER_KEY}/SET_ELEMENTS_COORDINATES`;
export const SET_ELEMENTS_SCORE_TO_CURRENT_SIZE = `${REDUCER_KEY}/SET_ELEMENTS_SCORE_TO_CURRENT_SIZE`;
export const SET_ELEMENTS_SIZES = `${REDUCER_KEY}/SET_ELEMENTS_SIZES`;
export const SET_ELEMENTS_SIZES_MIN_MAX = `${REDUCER_KEY}/SET_ELEMENTS_SIZES_MIN_MAX`;
export const TOGGLE_ASPECT_RATIO = `${REDUCER_KEY}/TOGGLE_ASPECT_RATIO`;
export const UPDATE_EVENTS_STATUS = `${REDUCER_KEY}/UPDATE_EVENTS_STATUS`;
export const UPDATE_PREV_STATE = `${REDUCER_KEY}/UPDATE_PREV_STATE`;
export const UNSELECT_ELEMENT = `${REDUCER_KEY}/UNSELECT_ELEMENT`;
