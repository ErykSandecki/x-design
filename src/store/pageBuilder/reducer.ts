// others
import {
  ADD_ELEMENT,
  APPLY_ELEMENTS_INSET_TYPE,
  APPLY_ELEMENTS_SIZE_MIN_MAX_TYPE,
  CHANGE_ALIGNMENT,
  CHANGE_BACKGROUND,
  CHANGE_LAYOUT,
  CHANGE_PARENT,
  CHANGE_POSITION,
  CLEAR_PREV_STATE,
  CHANGE_LAYOUT_BOX_SIZING,
  CHANGE_LAYOUT_GRID,
  CHANGE_INSETS,
  FIT_LAYOUT,
  FLIP_ELEMENTS,
  REDUCER_HISTORY_REDO,
  REDUCER_HISTORY_SAVE,
  REDUCER_HISTORY_UNDO,
  ROTATE_ELEMENTS,
  SELECT_ELEMENT,
  SELECT_ELEMENTS,
  SET_AREA_COORDINATES,
  RESIZE_ELEMENT,
  SET_ELEMENTS_COORDINATES,
  UNSELECT_ELEMENT,
  UPDATE_EVENTS_STATUS,
  UPDATE_PREV_STATE,
  SET_ELEMENTS_SIZES,
  APPLY_ELEMENTS_SIZE_TYPE,
  SET_ELEMENTS_SIZES_MIN_MAX,
  SET_ELEMENTS_SCORE_TO_CURRENT_SIZE,
  TOGGLE_ASPECT_RATIO,
  CHANGE_LAYOUT_ALIGNMENT,
  SET_ELEMENTS_GAP,
  APPLY_ELEMENTS_GAP_TYPE,
  CHANGE_CLIP_CONTENT,
} from './actionsType';
import { BASE_PAGE } from './constants';

// types
import { AnchorResize, AnchorRotate } from './enums';
import { KeyboardKeys, TAction } from 'types';
import {
  TAddELementAction,
  TSelectElementAction,
  TPageBuilderState,
  TUnselectElementAction,
  TSelectElementsAction,
  TSetAreaCoordinatesAction,
  TUpdateEventsStatusAction,
  TResizeElementAction,
  TRotateElementsAction,
  TChangeBackgroundAction,
  TReducerHistorySaveAction,
  TChangeAlignmentAction,
  TFlipElementsAction,
  TChangeLayoutAction,
  TSetElementsCoordinatesAction,
  TSetElementsSizesAction,
  TApplyElementsSizeTypeAction,
  TSetElementsSizesMinMaxAction,
  TSetElementsScoreToCurrentSizeAction,
  TChangeLayoutAlignmentAction,
  TSetElementsGapAction,
  TApplyElementsGapTypeAction,
  TChangeLayoutBoxSizingAction,
  TChangeLayoutGridAction,
  TChangeInsetsAction,
  TApplyElementsInsetTypeAction,
  TApplyElementsSizeMinMaxTypeAction,
  TChangeClipContentAction,
} from './types';

// utils
import { filterSelectedElements } from './utils/filterSelectedElements';
import { handleAddElement } from './utils/handleAddElement';
import { handleApplyElementsGapType } from './utils/applyElementsGapType/handleApplyElementsGapType';
import { handleApplyElementsInsetType } from './utils/applyElementsInsetType/handleApplyElementsInsetType';
import { handleApplyElementsSizeMinMaxType } from './utils/applyElementsSizeMinMaxType/handleApplyElementsSizeMinMaxType';
import { handleApplyElementsSizeType } from './utils/applyElementsSizeType/handleApplyElementsSizeType';
import { handleChangeAlignment } from './utils/changeAligment/handleChangeAlignment';
import { handleChangeBackground } from './utils/handleChangeBackground';
import { handleChangeClipContent } from './utils/handleChangeClipContent';
import { handleChangeInsets } from './utils/handleChangeInsets';
import { handleChangeLayout } from './utils/changeLayout/handleChangeLayout';
import { handleChangeLayoutAlignment } from './utils/handleChangeLayoutAlignment';
import { handleChangeLayoutBoxSizing } from './utils/handleChangeLayoutBoxSizing';
import { handleChangeLayoutGrid } from './utils/changeLayoutGrid/handleChangeLayoutGrid';
import { handleChangeParent } from './utils/changeParent/handleChangeParent';
import { handleChangePosition } from './utils/handleChangePosition';
import { handleFitLayout } from './utils/handleFitLayout';
import { handleFlipElements } from './utils/handleFlipElements';
import { handleReducerHistoryRedo } from './utils/reducerHistory/handleReducerHistoryRedo';
import { handleReducerHistorySave } from './utils/reducerHistory/handleReducerHistorySave';
import { handleReducerHistoryUndo } from './utils/reducerHistory/handleReducerHistoryUndo';
import { handleResizeElement } from './utils/resizeElement/handleResizeElement';
import { handleRotateElements } from './utils/handleRotateElements';
import { handleSetElementsCoordinates } from './utils/handleSetElementsCoordinates';
import { handleSetElementsGap } from './utils/handleSetElementsGap';
import { handleSetElementsScoreToCurrentSize } from './utils/handleSetElementsScoreToCurrentSize';
import { handleSetElementsSizes } from './utils/setElementSizes/handleSetElementsSizes';
import { handleSetElementsSizesMinMax } from './utils/handleSetElementsSizesMinMax';
import { handleToggleAspectRatio } from './utils/handleToggleAspectRatio';

const initialState: TPageBuilderState = {
  currentPage: '0',
  events: {
    canMoveElements: true,
    colorSampler: false,
    draggableElements: [],
    hoverOnElement: '-1',
    isGridDropArea: false,
    isMultipleMoving: false,
    isResizing: false,
    isRotating: false,
    possibleAnchorElementId: '-1',
    possibleAnchorPosition: null,
    possibleElement: undefined,
    possibleIndexPosition: null,
    possibleParent: null,
    pressedKey: KeyboardKeys.none,
    selectedAnchorResize: AnchorResize.none,
    selectedAnchorRotate: AnchorRotate.none,
  },
  isLoading: true,
  isPending: false,
  // @ts-ignore
  pages: {
    [BASE_PAGE.id]: {
      ...BASE_PAGE,
    },
  },
};

const addElement = (
  state: TPageBuilderState,
  { payload: element }: TAction<TAddELementAction['payload']>,
): TPageBuilderState => handleAddElement(element, state);

const applyElementsGapType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsGapTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsGapType(payload, state);

const applyElementsInsetType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsInsetTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsInsetType(payload, state);

const applyElementsSizeMinMaxType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsSizeMinMaxTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsSizeMinMaxType(payload, state);

const applyElementsSizeType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsSizeTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsSizeType(payload, state);

const changeAlignment = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeAlignmentAction['payload']>,
): TPageBuilderState => handleChangeAlignment(payload, state);

const changeBackground = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeBackgroundAction['payload']>,
): TPageBuilderState => handleChangeBackground(payload, state);

const changeClipContent = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeClipContentAction['payload']>,
): TPageBuilderState => handleChangeClipContent(payload, state);

const changeInsets = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeInsetsAction['payload']>,
): TPageBuilderState => handleChangeInsets(payload, state);

const changeLayout = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutAction['payload']>,
): TPageBuilderState => handleChangeLayout(payload, state);

const changeLayoutAlignment = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutAlignmentAction['payload']>,
): TPageBuilderState => handleChangeLayoutAlignment(payload, state);

const changeLayoutBoxSizing = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutBoxSizingAction['payload']>,
): TPageBuilderState => handleChangeLayoutBoxSizing(payload, state);

const changeLayoutGrid = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutGridAction['payload']>,
): TPageBuilderState => handleChangeLayoutGrid(payload, state);

const changeParent = (state: TPageBuilderState): TPageBuilderState => handleChangeParent(state);

const changePosition = (state: TPageBuilderState): TPageBuilderState => handleChangePosition(state);

const clearPrevState = (state: TPageBuilderState): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      prevState: undefined,
    },
  },
});

const fitLayout = (state: TPageBuilderState): TPageBuilderState => handleFitLayout(state);

const flipElements = (
  state: TPageBuilderState,
  { payload }: TAction<TFlipElementsAction['payload']>,
): TPageBuilderState => handleFlipElements(payload, state);

const reducerHistoryRedo = (state: TPageBuilderState): TPageBuilderState => handleReducerHistoryRedo(state);

const reducerHistorySave = (
  state: TPageBuilderState,
  { payload }: TAction<TReducerHistorySaveAction['payload']>,
): TPageBuilderState => handleReducerHistorySave(state, payload);

const reducerHistoryUndo = (state: TPageBuilderState): TPageBuilderState => handleReducerHistoryUndo(state);

const resizeElement = (
  state: TPageBuilderState,
  { payload: { baseCoordinates, flip, height, id, mouseCoordinates, width } }: TAction<TResizeElementAction['payload']>,
): TPageBuilderState => handleResizeElement(baseCoordinates, flip, height, width, id, mouseCoordinates, state);

const rotateElements = (
  state: TPageBuilderState,
  { payload }: TAction<TRotateElementsAction['payload']>,
): TPageBuilderState => handleRotateElements(payload, state);

const selectElement = (
  state: TPageBuilderState,
  { payload: selectedElement }: TAction<TSelectElementAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: filterSelectedElements(
        [...state.pages[state.currentPage].selectedElements, selectedElement],
        state,
      ),
    },
  },
});

const selectElements = (
  state: TPageBuilderState,
  { payload: selectedElements }: TAction<TSelectElementsAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: filterSelectedElements(selectedElements, state),
    },
  },
});

const setAreaCoordinates = (
  state: TPageBuilderState,
  { payload: areaCoordinates }: TAction<TSetAreaCoordinatesAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      areaCoordinates: {
        ...state.pages[state.currentPage].areaCoordinates,
        ...areaCoordinates,
      },
    },
  },
});

const setElementsCoordinates = (
  state: TPageBuilderState,
  { payload: coordinates }: TAction<TSetElementsCoordinatesAction['payload']>,
): TPageBuilderState => handleSetElementsCoordinates(coordinates, state);

const setElementsGap = (
  state: TPageBuilderState,
  { payload }: TAction<TSetElementsGapAction['payload']>,
): TPageBuilderState => handleSetElementsGap(payload, state);

const setElementsScoreToCurrentSize = (
  state: TPageBuilderState,
  { payload: { scoreType, sizeType } }: TAction<TSetElementsScoreToCurrentSizeAction['payload']>,
): TPageBuilderState => handleSetElementsScoreToCurrentSize(scoreType, sizeType, state);

const setElementsSizes = (
  state: TPageBuilderState,
  { payload: { sizeType, value } }: TAction<TSetElementsSizesAction['payload']>,
): TPageBuilderState => handleSetElementsSizes(sizeType, state, value);

const setElementsSizesMinMax = (
  state: TPageBuilderState,
  { payload: { scoreType, sizeType, value } }: TAction<TSetElementsSizesMinMaxAction['payload']>,
): TPageBuilderState => handleSetElementsSizesMinMax(scoreType, sizeType, state, value);

const toggleAspectRatio = (state: TPageBuilderState): TPageBuilderState => handleToggleAspectRatio(state);

const updateEventsStatus = (
  state: TPageBuilderState,
  { payload: events }: TAction<TUpdateEventsStatusAction['payload']>,
): TPageBuilderState => ({
  ...state,
  events: {
    ...state.events,
    ...events,
  },
});

const updatePrevState = (state: TPageBuilderState): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      prevState: state.pages[state.currentPage],
    },
  },
});

const unselectElement = (
  state: TPageBuilderState,
  { payload: id }: TAction<TUnselectElementAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: state.pages[state.currentPage].selectedElements.filter((element) => element.id !== id),
    },
  },
});

const pageBuilder = (state: TPageBuilderState = initialState, action: TAction): TPageBuilderState => {
  switch (action.type) {
    case ADD_ELEMENT:
      return addElement(state, action);
    case APPLY_ELEMENTS_GAP_TYPE:
      return applyElementsGapType(state, action);
    case APPLY_ELEMENTS_INSET_TYPE:
      return applyElementsInsetType(state, action);
    case APPLY_ELEMENTS_SIZE_MIN_MAX_TYPE:
      return applyElementsSizeMinMaxType(state, action);
    case APPLY_ELEMENTS_SIZE_TYPE:
      return applyElementsSizeType(state, action);
    case CHANGE_ALIGNMENT:
      return changeAlignment(state, action);
    case CHANGE_BACKGROUND:
      return changeBackground(state, action);
    case CHANGE_CLIP_CONTENT:
      return changeClipContent(state, action);
    case CHANGE_INSETS:
      return changeInsets(state, action);
    case CHANGE_LAYOUT:
      return changeLayout(state, action);
    case CHANGE_LAYOUT_ALIGNMENT:
      return changeLayoutAlignment(state, action);
    case CHANGE_LAYOUT_BOX_SIZING:
      return changeLayoutBoxSizing(state, action);
    case CHANGE_LAYOUT_GRID:
      return changeLayoutGrid(state, action);
    case CHANGE_PARENT:
      return changeParent(state);
    case CHANGE_POSITION:
      return changePosition(state);
    case CLEAR_PREV_STATE:
      return clearPrevState(state);
    case FIT_LAYOUT:
      return fitLayout(state);
    case FLIP_ELEMENTS:
      return flipElements(state, action);
    case REDUCER_HISTORY_REDO:
      return reducerHistoryRedo(state);
    case REDUCER_HISTORY_SAVE:
      return reducerHistorySave(state, action);
    case REDUCER_HISTORY_UNDO:
      return reducerHistoryUndo(state);
    case RESIZE_ELEMENT:
      return resizeElement(state, action);
    case ROTATE_ELEMENTS:
      return rotateElements(state, action);
    case SELECT_ELEMENT:
      return selectElement(state, action);
    case SELECT_ELEMENTS:
      return selectElements(state, action);
    case SET_AREA_COORDINATES:
      return setAreaCoordinates(state, action);
    case SET_ELEMENTS_COORDINATES:
      return setElementsCoordinates(state, action);
    case SET_ELEMENTS_GAP:
      return setElementsGap(state, action);
    case SET_ELEMENTS_SCORE_TO_CURRENT_SIZE:
      return setElementsScoreToCurrentSize(state, action);
    case SET_ELEMENTS_SIZES:
      return setElementsSizes(state, action);
    case SET_ELEMENTS_SIZES_MIN_MAX:
      return setElementsSizesMinMax(state, action);
    case TOGGLE_ASPECT_RATIO:
      return toggleAspectRatio(state);
    case UPDATE_EVENTS_STATUS:
      return updateEventsStatus(state, action);
    case UPDATE_PREV_STATE:
      return updatePrevState(state);
    case UNSELECT_ELEMENT:
      return unselectElement(state, action);
    default:
      return state;
  }
};

export default pageBuilder;
