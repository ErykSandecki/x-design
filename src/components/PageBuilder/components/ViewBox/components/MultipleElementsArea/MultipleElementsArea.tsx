import { createPortal } from 'react-dom';
import { FC } from 'react';
import { useSelector } from 'react-redux';

// components
import ClickableArea from './components/ClickableArea/ClickableArea';
import Corners from '../Corners/Corners';

// core
import { useRefs } from 'pages/PageBuilderPage/core/RefsProvider';

// store
import { multipleSelectedElementsSelector } from 'store/pageBuilder/selectors';

// utils
import { useMultipleELementsAreaEvents } from './hooks/useMultipleELementsAreaEvents';

const MultipleElementsArea: FC = () => {
  const isMultiple = useSelector(multipleSelectedElementsSelector);
  const { overlayContainerRef } = useRefs();
  const { coordinatesData, showCorners } = useMultipleELementsAreaEvents();

  if (!isMultiple) {
    return null;
  }

  return (
    <>
      <ClickableArea
        elementsCordinates={coordinatesData.elementsCoordinates}
        outlineCoordinates={coordinatesData.outline}
      />
      {showCorners && createPortal(<Corners rectCoordinates={coordinatesData.outline} />, overlayContainerRef.current)}
    </>
  );
};

export default MultipleElementsArea;
