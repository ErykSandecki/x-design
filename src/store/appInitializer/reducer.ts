// others
import { DEFAULT_LANGUAGE } from 'translations';
import { INIT_LANGUAGE_SUCCESS, SET_IS_APP_LOADED, SET_THEME } from './actionsType';

// types
import { TAction } from 'types';
import { TAppInitializerState, TInitLanguageSuccessAction, TSetIsAppLoadedAction, TSetThemeAction } from './types';

// utils
import { getThemePreferences } from 'utils';

const initialState: TAppInitializerState = {
  isAppLoaded: false,
  isPending: true,
  language: DEFAULT_LANGUAGE,
  theme: getThemePreferences(),
};

const initLanguageSuccess = (
  state: TAppInitializerState,
  { payload: language }: TAction<TInitLanguageSuccessAction['payload']>,
): TAppInitializerState => ({
  ...state,
  language,
});

const setIsAppLoaded = (
  state: TAppInitializerState,
  { payload: isAppLoaded }: TAction<TSetIsAppLoadedAction['payload']>,
): TAppInitializerState => ({
  ...state,
  isAppLoaded,
  isPending: false,
});

const setTheme = (
  state: TAppInitializerState,
  { payload: theme }: TAction<TSetThemeAction['payload']>,
): TAppInitializerState => ({
  ...state,
  theme,
});

const appInitializer = (state: TAppInitializerState = initialState, action: TAction): TAppInitializerState => {
  switch (action.type) {
    case INIT_LANGUAGE_SUCCESS:
      return initLanguageSuccess(state, action);
    case SET_IS_APP_LOADED:
      return setIsAppLoaded(state, action);
    case SET_THEME:
      return setTheme(state, action);
    default:
      return state;
  }
};

export default appInitializer;
