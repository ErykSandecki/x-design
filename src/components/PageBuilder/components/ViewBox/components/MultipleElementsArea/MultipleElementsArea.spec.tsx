import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import MultipleElementsArea from './MultipleElementsArea';

// core
import { RefsProvider } from 'pages/PageBuilderPage/core/RefsProvider';

// mocks
import { elementMock, pageBuilderStateMock, selectedElementMock } from 'test/mocks/reducer/pageBuilderMock';
import { wholeStateMock } from 'test/mocks/reducer/wholeStateMock';

// others
import { REDUCER_KEY as PAGE_BUILDER } from 'store/pageBuilder/actionsType';

// store
import { configureStore, store as storeToMock } from 'store/store';

const element = document.createElement('div');
const currentPage = pageBuilderStateMock[PAGE_BUILDER].pages[pageBuilderStateMock[PAGE_BUILDER].currentPage];
const overlayContainer = document.createElement('div');
const zoomContent = document.createElement('div');

const sharedRefs = {
  [elementMock.id]: element,
  ['2']: element,
};

const stateMock = {
  [PAGE_BUILDER]: {
    ...pageBuilderStateMock[PAGE_BUILDER],
    pages: {
      ...pageBuilderStateMock[PAGE_BUILDER].pages,
      ['0']: {
        ...pageBuilderStateMock[PAGE_BUILDER].pages['0'],
        elements: {
          ...currentPage.elements,
          allData: {
            ...currentPage.elements.allData,
            [elementMock.id]: {
              ...elementMock,
            },
            ['2']: {
              ...elementMock,
              id: '2',
            },
          },
        },
        selectedElements: [
          selectedElementMock,
          {
            ...selectedElementMock,
            id: '2',
          },
        ],
      },
    },
  },
};

jest.mock('lodash', () => ({
  ...(jest.requireActual('lodash') as object),
  defer: (callback: any): any => callback(),
}));

describe('MultipleElementsArea snapshots', () => {
  beforeAll(() => {
    // mock
    element.style.height = '100px';
    element.style.width = '100px';
    zoomContent.style.height = '1000px';
    zoomContent.style.width = '1000px';
    document.body.appendChild(zoomContent);
    document.body.appendChild(overlayContainer);
    storeToMock.getState = (): any =>
      ({
        ...wholeStateMock,
        [PAGE_BUILDER]: {
          ...pageBuilderStateMock[PAGE_BUILDER],
          pages: {
            ...pageBuilderStateMock[PAGE_BUILDER].pages,
            [currentPage.id]: {
              ...currentPage,
              elements: {
                ...currentPage.elements,
                allData: {
                  ...currentPage.elements.allData,
                  [selectedElementMock.id]: {
                    ...elementMock,
                  },
                  ['2']: {
                    ...elementMock,
                    id: '2',
                  },
                },
              },
            },
          },
        },
      }) as any;
  });

  it('should render MultipleElementsArea', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <RefsProvider
          itemsRefs={sharedRefs}
          overlayContainerRefHtml={overlayContainer}
          zoomContentRefHtml={zoomContent}
        >
          <MultipleElementsArea />
        </RefsProvider>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not render MultipleElementsArea', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <RefsProvider
          itemsRefs={sharedRefs}
          overlayContainerRefHtml={overlayContainer}
          zoomContentRefHtml={zoomContent}
        >
          <MultipleElementsArea />
        </RefsProvider>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
