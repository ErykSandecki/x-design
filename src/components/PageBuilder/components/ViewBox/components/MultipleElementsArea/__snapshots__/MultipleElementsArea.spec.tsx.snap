// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleElementsArea snapshots should not render MultipleElementsArea 1`] = `<DocumentFragment />`;

exports[`MultipleElementsArea snapshots should render MultipleElementsArea 1`] = `
<DocumentFragment>
  <svg
    class="ClickableArea ClickableArea--light"
    data-test-outline=""
    style="height: 101.5px; left: -0.75px; top: -0.75px; width: 101.5px;"
  >
    <path
      class="ClickableArea__outline ClickableArea__outline--light"
      d="M 0,0 H 101.5 V 101.5 H 0 V 0"
      stroke-width="1.5"
    />
  </svg>
  <svg
    class="ClickableArea ClickableArea--light"
    data-test-rect="test-1"
    style="height: 100px; left: 0px; pointer-events: all; top: 0px; width: 100px;"
  >
    <rect
      fill="none"
      height="100"
      width="100"
      x="0"
      y="0"
    />
  </svg>
  <svg
    class="ClickableArea ClickableArea--light"
    data-test-rect="2"
    style="height: 100px; left: 0px; pointer-events: all; top: 0px; width: 100px;"
  >
    <rect
      fill="none"
      height="100"
      width="100"
      x="0"
      y="0"
    />
  </svg>
  <svg
    class="Corners Corners--light"
    style="height: 101.5px; left: -0.75px; top: -0.75px; width: 101.5px;"
  >
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="-3"
      y="-3"
    />
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="-3"
      y="97"
    />
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="97"
      y="-3"
    />
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="97"
      y="97"
    />
  </svg>
</DocumentFragment>
`;
